File: src/scanner.test.ts
import fs from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import { spawnSync } from "node:child_process";
import { test } from "node:test";
import assert from "node:assert/strict";
import { findReposWithNodeModules } from "./scanner.js";

const run = (args: string[], cwd: string) => {
  const result = spawnSync("git", args, { cwd, encoding: "utf8" });
  if (result.status !== 0) {
    throw new Error(result.stderr || result.stdout || `git ${args.join(" ")} failed`);
  }
};

const writeRepo = async (
  root: string,
  name: string,
  options: { lockfiles?: string[]; commitDate?: string } = {}
): Promise<string> => {
  const repoPath = path.join(root, name);
  await fs.mkdir(path.join(repoPath, "node_modules"), { recursive: true });
  await fs.writeFile(path.join(repoPath, "package.json"), "{}");
  for (const lockfile of options.lockfiles ?? []) {
    await fs.writeFile(path.join(repoPath, lockfile), "");
  }
  await fs.writeFile(path.join(repoPath, "node_modules", "file.txt"), "x");

  run(["init"], repoPath);
  run(["add", "."], repoPath);
  run(["config", "user.email", "test@example.com"], repoPath);
  run(["config", "user.name", "Test User"], repoPath);
  const env = {
    ...process.env,
    ...(options.commitDate ? { GIT_AUTHOR_DATE: options.commitDate, GIT_COMMITTER_DATE: options.commitDate } : {})
  };
  const result = spawnSync("git", ["commit", "-m", "init"], { cwd: repoPath, encoding: "utf8", env });
  if (result.status !== 0) {
    throw new Error(result.stderr || result.stdout || "git commit failed");
  }

  return repoPath;
};

test("findReposWithNodeModules detects package manager from lockfile priority", async () => {
  const tempRoot = await fs.mkdtemp(path.join(os.tmpdir(), "repo-cleanup-test-"));
  try {
    await writeRepo(tempRoot, "repo-multi-lock", { lockfiles: ["yarn.lock", "package-lock.json"] });

    const rows = await findReposWithNodeModules(tempRoot);
    assert.equal(rows.length, 1);
    assert.equal(rows[0]?.manager, "yarn");
    assert.equal(rows[0]?.hasLockfile, true);
    assert.equal(rows[0]?.reinstallCommand, "yarn install --immutable");
    assert.ok((rows[0]?.git.branch?.length ?? 0) > 0);
    assert.equal(rows[0]?.git.ahead, 0);
    assert.equal(rows[0]?.git.behind, 0);
  } finally {
    await fs.rm(tempRoot, { recursive: true, force: true });
  }
});

test("findReposWithNodeModules computes inactivity days from git history", async () => {
  const tempRoot = await fs.mkdtemp(path.join(os.tmpdir(), "repo-cleanup-test-"));
  try {
    await writeRepo(tempRoot, "repo-old", { lockfiles: ["pnpm-lock.yaml"], commitDate: "2020-01-01T00:00:00Z" });
    const rows = await findReposWithNodeModules(tempRoot);
    assert.equal(rows.length, 1);
    assert.equal(rows[0]?.manager, "pnpm");
    assert.ok((rows[0]?.inactiveDays ?? 0) > 365, "inactiveDays should reflect old commit");
  } finally {
    await fs.rm(tempRoot, { recursive: true, force: true });
  }
});

test("findReposWithNodeModules reports scan progress", async () => {
  const tempRoot = await fs.mkdtemp(path.join(os.tmpdir(), "repo-cleanup-test-"));
  try {
    await writeRepo(tempRoot, "repo-progress", { lockfiles: ["yarn.lock"] });
    let progressEvents = 0;
    let lastCount = 0;
    await findReposWithNodeModules(tempRoot, {
      onProgress: ({ directoriesScanned }) => {
        progressEvents += 1;
        lastCount = directoriesScanned;
      }
    });
    assert.ok(progressEvents > 0);
    assert.ok(lastCount > 0);
  } finally {
    await fs.rm(tempRoot, { recursive: true, force: true });
  }
});

test("findReposWithNodeModules emits rows as they are found", async () => {
  const tempRoot = await fs.mkdtemp(path.join(os.tmpdir(), "repo-cleanup-test-"));
  try {
    const repoA = await writeRepo(tempRoot, "repo-a", { lockfiles: ["yarn.lock"] });
    const repoB = await writeRepo(tempRoot, "repo-b", { lockfiles: ["pnpm-lock.yaml"] });
    const found: string[] = [];
    const rows = await findReposWithNodeModules(tempRoot, {
      onRepoFound: (row) => found.push(row.repoPath)
    });

    assert.equal(rows.length, 2);
    assert.equal(found.length, 2);
    assert.ok(found.includes(repoA));
    assert.ok(found.includes(repoB));
  } finally {
    await fs.rm(tempRoot, { recursive: true, force: true });
  }
});

test("findReposWithNodeModules emits size updates for discovered rows", async () => {
  const tempRoot = await fs.mkdtemp(path.join(os.tmpdir(), "repo-cleanup-test-"));
  try {
    await writeRepo(tempRoot, "repo-size-update", { lockfiles: ["yarn.lock"] });
    const updates: number[] = [];
    const rows = await findReposWithNodeModules(tempRoot, {
      onRepoUpdated: (row) => updates.push(row.bytes)
    });

    assert.equal(rows.length, 1);
    assert.ok(updates.length >= 1);
    assert.ok(updates.some((bytes) => bytes > 0));
    assert.ok((rows[0]?.bytes ?? 0) > 0);
  } finally {
    await fs.rm(tempRoot, { recursive: true, force: true });
  }
});

File: internal/scanner/scanner_test.go
package scanner

import (
	"os"
	"os/exec"
	"path/filepath"
	"sync"
	"testing"
)

func runGit(t *testing.T, dir string, args ...string) {
	t.Helper()
	cmd := exec.Command("git", args...)
	cmd.Dir = dir
	out, err := cmd.CombinedOutput()
	if err != nil {
		t.Fatalf("git %v: %s %v", args, out, err)
	}
}

func writeRepo(t *testing.T, root, name string, lockfiles []string, commitDate string) string {
	t.Helper()
	repo := filepath.Join(root, name)
	if err := os.MkdirAll(filepath.Join(repo, "node_modules"), 0755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(filepath.Join(repo, "package.json"), []byte("{}"), 0644); err != nil {
		t.Fatal(err)
	}
	for _, lf := range lockfiles {
		if err := os.WriteFile(filepath.Join(repo, lf), nil, 0644); err != nil {
			t.Fatal(err)
		}
	}
	if err := os.WriteFile(filepath.Join(repo, "node_modules", "file.txt"), []byte("x"), 0644); err != nil {
		t.Fatal(err)
	}
	runGit(t, repo, "init")
	runGit(t, repo, "add", ".")
	runGit(t, repo, "config", "user.email", "test@example.com")
	runGit(t, repo, "config", "user.name", "Test User")
	cmd := exec.Command("git", "commit", "-m", "init")
	cmd.Dir = repo
	if commitDate != "" {
		cmd.Env = append(os.Environ(), "GIT_AUTHOR_DATE="+commitDate, "GIT_COMMITTER_DATE="+commitDate)
	}
	if out, err := cmd.CombinedOutput(); err != nil {
		t.Fatalf("git commit: %s %v", out, err)
	}
	return repo
}

func TestDetectManagerLockfilePriority(t *testing.T) {
	root := t.TempDir()
	writeRepo(t, root, "multi", []string{"yarn.lock", "package-lock.json"}, "")

	rows, err := Scan(root, Options{})
	if err != nil {
		t.Fatal(err)
	}
	if len(rows) != 1 {
		t.Fatalf("rows: %d", len(rows))
	}
	if rows[0].Manager != ManagerYarn || !rows[0].HasLockfile {
		t.Fatalf("manager: %+v", rows[0])
	}
	if rows[0].ReinstallCommand != "yarn install --immutable" {
		t.Fatalf("restore: %q", rows[0].ReinstallCommand)
	}
	if rows[0].Git.Branch == "" {
		t.Fatal("expected branch name")
	}
	if rows[0].Git.Ahead != 0 || rows[0].Git.Behind != 0 {
		t.Fatalf("ahead/behind: %+v", rows[0].Git)
	}
	if rows[0].Bytes <= 0 {
		t.Fatal("expected bytes > 0")
	}
}

func TestInactiveDaysFromCommit(t *testing.T) {
	root := t.TempDir()
	writeRepo(t, root, "old", []string{"pnpm-lock.yaml"}, "2020-01-01T00:00:00Z")

	rows, err := Scan(root, Options{})
	if err != nil {
		t.Fatal(err)
	}
	if len(rows) != 1 {
		t.Fatalf("rows: %d", len(rows))
	}
	if rows[0].Manager != ManagerPnpm {
		t.Fatalf("manager: %q", rows[0].Manager)
	}
	if rows[0].InactiveDays == nil || *rows[0].InactiveDays < 365 {
		t.Fatalf("inactiveDays: %v", rows[0].InactiveDays)
	}
}

func TestScanProgress(t *testing.T) {
	root := t.TempDir()
	writeRepo(t, root, "prog", []string{"yarn.lock"}, "")

	var events int
	_, err := Scan(root, Options{
		OnProgress: func(p Progress) {
			if p.DirectoriesScanned > 0 {
				events++
			}
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	if events == 0 {
		t.Fatal("expected progress events")
	}
}

func TestScanFoundAndSizedCallbacks(t *testing.T) {
	root := t.TempDir()
	repoA := writeRepo(t, root, "a", []string{"yarn.lock"}, "")
	repoB := writeRepo(t, root, "b", []string{"pnpm-lock.yaml"}, "")

	var (
		mu    sync.Mutex
		found []string
		sized []int64
	)
	rows, err := Scan(root, Options{
		OnFound: func(c Candidate) {
			mu.Lock()
			found = append(found, c.RepoPath)
			mu.Unlock()
		},
		OnSized: func(c Candidate) {
			mu.Lock()
			sized = append(sized, c.Bytes)
			mu.Unlock()
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	if len(rows) != 2 || len(found) != 2 {
		t.Fatalf("rows=%d found=%d", len(rows), len(found))
	}
	if !contains(found, repoA) || !contains(found, repoB) {
		t.Fatalf("found: %v", found)
	}
	if len(sized) < 2 {
		t.Fatalf("sized: %v", sized)
	}
	for _, b := range sized {
		if b <= 0 {
			t.Fatal("expected positive size")
		}
	}
}

func contains(ss []string, s string) bool {
	for _, v := range ss {
		if v == s {
			return true
		}
	}
	return false
}
